packages/playwright-core/src/server/page.ts

```diff
@@ -846,54 +849,66 @@ function addPageBinding(bindingName: string, needsHandle: boolean, utilityScript
 
 class FrameThrottler {
   private _acks: (() => void)[] = [];
-  private _interval: number;
+  private _defaultInterval: number;
+  private _throttlingInterval: number;
   private _nonThrottledFrames: number;
   private _budget: number;
-  private _intervalId: NodeJS.Timeout | undefined;
+  private _throttlingEnabled = false;
+  private _timeoutId: NodeJS.Timeout | undefined;
 
-  constructor(nonThrottledFrames: number, interval: number) {
+  constructor(nonThrottledFrames: number, defaultInterval: number, throttlingInterval: number) {
     this._nonThrottledFrames = nonThrottledFrames;
     this._budget = nonThrottledFrames;
-    this._interval = interval;
+    this._defaultInterval = defaultInterval;
+    this._throttlingInterval = throttlingInterval;
+    this._tick();
   }
 
-  setEnabled(enabled: boolean) {
-    if (enabled) {
-      if (this._intervalId)
-        clearInterval(this._intervalId);
-      this._intervalId = setInterval(() => this._tick(), this._interval);
-    } else if (this._intervalId) {
-      clearInterval(this._intervalId);
-      this._intervalId = undefined;
+  dispose() {
+    if (this._timeoutId) {
+      clearTimeout(this._timeoutId);
+      this._timeoutId = undefined;
     }
   }
 
+  setThrottlingEnabled(enabled: boolean) {
+    this._throttlingEnabled = enabled;
+  }
+
   recharge() {
     // Send all acks, reset budget.
     for (const ack of this._acks)
       ack();
     this._acks = [];
     this._budget = this._nonThrottledFrames;
+    if (this._timeoutId) {
+      clearTimeout(this._timeoutId);
+      this._tick();
+    }
   }
 
   ack(ack: () => void) {
-    // Either not engaged or video is also recording, don't throttle.
-    if (!this._intervalId) {
-      ack();
-      return;
-    }
-
-    // Do we have enough budget to respond w/o throttling?
-    if (--this._budget > 0) {
+    if (!this._timeoutId) {
+      // Already disposed.
       ack();
       return;
     }
-
-    // Schedule.
     this._acks.push(ack);
   }
 
   private _tick() {
-    this._acks.shift()?.();
+    const ack = this._acks.shift();
+    if (ack) {
+      --this._budget;
+      ack();
+    }
+
+    if (this._throttlingEnabled && this._budget <= 0) {
+      // Non-throttled frame budget is exceeded. Next ack will be throttled.
+      this._timeoutId = setTimeout(() => this._tick(), this._throttlingInterval);
+    } else {
+      // Either not throttling, or still under budget. Next ack will be after the default timeout.
+      this._timeoutId = setTimeout(() => this._tick(), this._defaultInterval);
```

Comment:
should it be something like Math.max(0, this._defaultTimeout - this._timeSinceLastAck) ?